### PR TITLE
Overlapping in drawio

### PR DESCRIPTION
**Pull request by haim-kermany**
overlapping

### CODE REVIEW

File: pkg/drawio/layoutOverlap.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+			continue
+		}
+		line := tn.(LineTreeNodeInterface)
+		newLinePoint := []point{}
+		oldLinePoints := line.Points()
+		absPoints := getLineAbsolutePoints(line)
+		for pointIndex := range absPoints[0 : len(absPoints)-1] {
+			srcP := absPoints[pointIndex]
+			desP := absPoints[pointIndex+1]
+			for {
```
Comment (zivnevo) [inline pkg/drawio/layoutOverlap.go:123]:
```suggestion
			for { // add bypass points until there is no overlap with icons (or until no bypass can be found)
```

**Review by zivnevo — CHANGES_REQUESTED** [review]
Great code!
See some comments.

File: pkg/drawio/layoutOverlap.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+		return a
+	}
+	return -a
+}
+
+func pow(a, b int) int {
+	if b <= 0 {
+		return 1
+	}
+	return pow(a, b-1) * a
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go:266]: done

File: pkg/drawio/layoutOverlap.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+// in both steps 2,3. choosing the bypass point is done in two steps:
+//    1. calculate a list of potential bypass points, around the icon, (or around the middle of the line)
+//    2. choose the best point, and add it to the line.
+
+var NOPOINT = point{-1, -1}
+
+const nPotentialBP = 6
+
+type overlapCell struct {
+	hasBP bool
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go:31]: done

File: pkg/drawio/layoutOverlap.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+// Handling overlapping is done in three steps:
+//    1. update the matrix with the icons locations
+//    2. handling pairs of lines sharing the same path -
+//       this is done by adding a bypass point to one of the lines, at the middle of the line
+//    3. handling lines overlapping  icons - this is done by adding a bypass points to the line, next to the icons
+// in both steps 2,3. choosing the bypass point is done in two steps:
+//    1. calculate a list of potential bypass points, around the icon, (or around the middle of the line)
+//    2. choose the best point, and add it to the line.
+
+var NOPOINT = point{-1, -1}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go:26]: done

File: pkg/drawio/layoutOverlap.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+	for i := 0; i < nPotentialBP; i++ {
+		verticalVectorSize := pow(-1, i) * (1 + i/2) * iconSize
+		verticalVectorX := verticalVectorSize * deltaY / disXY
+		verticalVectorY := verticalVectorSize * deltaX / disXY
+		BP := point{max(0, middlePoint.X+verticalVectorX), max(0, middlePoint.Y-verticalVectorY)}
+		if lyO.cell(BP.X, BP.Y).icon != nil {
+			continue
+		}
+		for lyO.cell(BP.X, BP.Y).hasBP {
+			BP = point{BP.X + minSize, BP.Y + minSize}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go:165]: fixed

File: pkg/drawio/layoutOverlap.go [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+			return icon
+		}
+	}
+	return nil
+}
+
+// some methods to convert absolute point to relative, and vis versa:
+func getLineAbsolutePoints(line LineTreeNodeInterface) []point {
+	absPoints := []point{}
+	absPoints = append(absPoints, iconCenterPoint(line.Src()))
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go:215]: done

File: pkg/drawio/layoutOverlap.go [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+		}
+		for lyO.cell(BP.X, BP.Y).hasBP {
+			BP = point{BP.X + minSize, BP.Y + minSize}
+		}
+		BPs = append(BPs, BP)
+	}
+	return BPs
+}
+
+// getBypassPoint() select the best BS, the closest the has minimal overlapping
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go:172]: fixed

File: pkg/drawio/layoutOverlap.go [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+			lyO.cell(BP.X, BP.Y).hasBP = true
+			return BP
+		}
+	}
+	for _, BP := range BPs {
+		if lyO.getOverlappedIcon(BP, dstPoint, line) == nil {
+			lyO.cell(BP.X, BP.Y).hasBP = true
+			return BP
+		}
+	}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go:192]: yes.
after choosing the BP, we handle the BP->dst interval. so if there is an overlap on BP->dst interval, we will solve it later.
but overlap in src->BP will stay there.

[the remark this answers is not in the document]
Comment (zivnevo) [reply pkg/drawio/layoutOverlap.go]: ok, got it.

File: pkg/drawio/treeNodeInterface.go [hunk 9]
```diff
@@ -96,3 +96,10 @@ func locations(tns []TreeNodeInterface) []*Location {
 	}
 	return locations
 }
+
+func absoluteGeometry(tn TreeNodeInterface) (x, y int) {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/treeNodeInterface.go:100]: kacha

File: pkg/drawio/layoutOverlap.go [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+// handleLinesOverLines() - find pairs of overlapping lines, and add point to one of them
+func (lyO *layoutOverlap) handleLinesOverLines() {
+	nodes := getAllNodes(lyO.network)
+	for i1 := range nodes {
+		for i2 := i1 + 1; i2 < len(nodes); i2++ {
+			if !nodes[i1].IsLine() || !nodes[i2].IsLine() {
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go]: fixed

**Review by zivnevo — DISMISSED** [review]

**Review by zivnevo — APPROVED** [review]

**Pull request merged**